DO NOT MERGE: LwM2M holding PR for 1.10

mike-scott opened the pull request:
Currently, the LwM2M uses net_context structures and the LwM2M sample has code to establish network communications.

This should be changed to use the net_app APIs so that the sample can remove most of the network
setup code and the library itself can benefit from other net_app features such as DTLS, etc.

rbtchc committed:
net: lwm2m: add state machine for firmware pull update

1. Parse firmware pull URI
2. Add lwm2m_firmware_get/set_update_cb() for applicaiton to register
   callback. This is because we want to check the update_state before
   we pass to the application
3. Add lwm2m_firmware_get/set_update_result() and
   lwm2m_firmware_get/set_update_stat() to manage the state trasnition
   as well as the sanity check

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>
[michael.scott@linaro.org: rebased on net_app framework and
lwm2m_message refactoring.]
Signed-off-by: Michael Scott <michael.scott@linaro.org>

rbtchc committed:
net: lwm2m: add firmware push support

1. Add handling block1 option in handle_request(). The basic idea is
   to declare structure block_context at compiled time and use "token"
   as a key to pick up the on-going block cotext. It should be able to
   support multiple blockwise transfer concurrently
2. Use write callback implemented in lwm2m_obj_firmware to deal w/ the
   update state transition and than call the callback registered by the
   application
3. move default_block_size to lwm2m_engine.c to share btw lwm2m_engine
   and lwm2m_obj_firmware_pull

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>
[michael.scott@linaro.org: rebased on LwM2M net_app changes.]
Signed-off-by: Michael Scott <michael.scott@linaro.org>

rsalveti committed:
net: lwm2m: firmware: add support for firmware pull over CoAP proxy

CoAP allows a proxy to be used when transferring data (CoAP-CoAP and/or
CoAP-HTTP) by creating request on a specific URI path and by using the
Proxy URI CoAP option. Create specific Kconfig options for the proxy
server address and port, until a parser gets implemented.

Code tested with Californium acting as CoAP proxy.

Signed-off-by: Ricardo Salveti <ricardo.salveti@linaro.org>
[michael.scott@linaro.org: rebased on net_app + lwm2m_message refactoring
+ firmware update changes.]
Signed-off-by: Michael Scott <michael.scott@linaro.org>

rbtchc committed:
net: http: add config to separate URL only parser

Add HTTP_PARSER_URL_ONLY to separate URL only parser

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc committed:
net: lwm2m: setup data_ptr/len for OPAQUE resource when none given

OPAQUE resource type might/might not have data_ptr/data_len setup
depending on the implementation. This introduce an issue that when
OPAQUE resource is written from the server side, the ones w/ none
setup will not be able to get the data at post_write_cb()

Modify to setup data_ptr/data_len as incoming buffer and buffer size

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rveerama1 committed:
net: samples: Add leds demo app over RPL and CoAP

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

mike-scott commented:
@rbtchc I realize this code will conflict with your PR: https://github.com/zephyrproject-rtos/zephyr/pull/1049  We'll need to work out which will be merged first and how. :/  Please take a look.

mike-scott commented:
@pfalcon @rsalveti Please take a look as well.

jukkar reviewed (commented):
Lot of changes, and I probably missed something. Anyway, this is a good start.

mike-scott commented on include/net/lwm2m.h:
```diff
@@ -25,6 +25,16 @@
+ * @details Context structure for the LwM2M high-level API.
+ */
+struct lwm2m_ctx {
+	/** Net app context structure */
```
Woops.. by the end of the patch set this is a net_app_ctx.. I probably didn't adjust the comment mid-patch series.  I'll fix.

mike-scott commented on samples/net/lwm2m_client/src/lwm2m-client.c:
```diff
@@ -57,10 +57,12 @@ static struct device *led_dev;
 
 #if defined(CONFIG_NET_IPV6)
-static struct net_app_ctx udp6;
+static struct net_app_ctx app_udp6;
```
Hrm.  At the end of the patch series, I end up with 2 lwm2m_ctx structures (each with it's own net_app_ctx).  They handle IPv4 and IPv6 individually.  Do we have any samples which illustrate dual use of a net_app_ctx?

jukkar replied:
Hmm, I seem to have talked about server side but the lwm2m is about client. So in client side this is not possible, sorry for confusion.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2174,15 +2175,17 @@ static int handle_request(struct zoap_packet *request,
 void lwm2m_udp_receive(struct lwm2m_ctx *client_ctx, struct net_pkt *pkt,
 		       bool handle_separate_response,
-		       int (*udp_request_handler)(struct zoap_packet *,
+		       int (*udp_request_handler)(struct net_context *net_ctx,
```
Will do.  I think early on I was heeding the checkpatch warning about no new typedefs :P

mike-scott commented:
Grr.. Github still hiding comments as "outdated"..  annoying.

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -554,15 +547,15 @@ int lwm2m_init_message(struct net_context *net_ctx, struct zoap_packet *zpkt,
-						int num_zpendings)
+struct zoap_pending *lwm2m_init_message_pending(struct lwm2m_ctx *client_ctx,
+						struct zoap_packet *zpkt,
+						struct sockaddr *addr)
```
I suppose "addr" can be removed. Peer sockaddr can be obtained from client_ctx->net_app_ctx->default_ctx->remote

mike-scott replied:
Eventually the addr here will be resolved from the security objects (see page 121 of the LwM2M spec example URL: coaps://bootstrap.example.com).   If the 0/0/0 resource is a bootstrap server it will then point us to another server which would need to be resolved, etc.

mike-scott replied:
@rbtchc Your original comment here now applies.  Going forward the rd_client will start engine with peer_addr information on a per server basis.  So net_app_ctx.default_ctx->remote will be the valid destination address as far as the engine is concerned.

I will add a commit to V3 addressing this.

mike-scott replied:
I also retract my earlier agreement.  As with lwm2m_udp_sendto(), this function is used by the observer notify logic.  And there shouldn't be an inherent limitation as to who can subscribe to notifications except where access controls (to be added later) allow.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2181,17 +2175,17 @@ static int handle_request(struct zoap_packet *request,
 
-int lwm2m_udp_sendto(struct net_pkt *pkt, const struct sockaddr *dst_addr)
+int lwm2m_udp_sendto(struct net_app_ctx *app_ctx, struct net_pkt *pkt,
+		     const struct sockaddr *dst_addr)
```
Same as above.

mike-scott replied:
And the same here.

mike-scott replied:
Actually, I take it back.  lwm2m_udp_sendto() is used to send to *any* address which may have sent a message to the IoT device (not just the peer registered in the net_app_ctx).  For instance, if you make a query via Copper plug-in, that IP address needs to be used in lwm2m_udp_sendto().

That being said, this brings up a security question.  And there will be more patches in the future to add the access controls object.  This should stop production devices from responding to anyone not allowed.

rbtchc replied:
Agree. But I suppose it doesn't work when DTLS is enabled.

rbtchc commented:
I check the connections on the shell console by executing command
1. select net
2. conn
And it prompts
>      Context   	        Iface                Flags Local           	                Remote
>      [ 1] 0x00405cc0	0x00405220    4DU   AF_UNSPEC	
>      [ 2] 0x00405d1c	0x00405220    6DU   [2001:db8::1]:34161	[2001:db8::2]:5683
>      [ 3] 0x00405d78	0x00405220    4DU   192.0.2.1:37773	        192.0.2.2:5683
>      [ 4] 0x00405dd4	0x00405220    6DU   AF_UNSPEC	

I'm not sure why there are four connections existing on the list.
I suppose there should only be two of them one for IPv4 and one for IPv6 clients.

I, too, alter the interfaces to net_app while I was working on the bootstrap interfaces.
But I don't have the connection w/ Local equals to AF_UNSPEC.
I will try to figure out why.

mike-scott commented:
@rbtchc  Please don't test this too much atm.  I have 2 net_app_ctx and I'm going to remove 1.  That's the reason we have 2 extra AF_UNSPEC in the conn list.

I'll submit a new PR tomorrow during my daytime.

rbtchc commented:
Right. I figure out there are two extra net_app_ctx at lwm2m_obj_firmware_pull.c

Commit:
net: lwm2m: add support for DTLS

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: lwm2m: add support for DTLS

For QEMU testing use CONF_FILE=prj_qemu_x86_dtls.conf
For K64F testing use CONF_FILE=prj_frdm_k64f_dtls.conf

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: set range for client lifetime config

Due to timeout checking the minimum lifetime must be 15 seconds,
and we're storing the lifetime as an unsigned short so set the
maximum to 65535.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

rveerama1 committed:
net: samples: Modify coaps_server to new CoAP api's

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: samples: Modify coaps_client to new CoAP api's

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: coap: Rename ZOAP library

ZOAP library has certain limitations in parsing and preparation of
coap messages. It can handle only on single network fragment. If
network packet is split between multiple fragments it fails. This
patch is just copy and rename of 'zoap' to 'coap'.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: coap: Add multi fragment support to CoAP library

Current coap library fails to parse or prepare if packet is more
than one fragment. Added support to handle multi fragment packet.
Also well-known/core api used to prepare coap packet and send it
through net context api immediately. This is goind to be problematic
if user doesn't enable net context. Also user can not encrypt coap
packets. Now api will return prepared coap packet to application.
Application will send it to peer.

Jira: ZEP-2210

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: tests: Update CoAP tests to use new API

ZoAP tests are modified to use new CoAP API. Also modified tests
name from 'zoap' to 'coap'.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: samples: Modify leds_demo to use new CoAP api's

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: samples: Modify zoap_server to new CoAP api's

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: samples: Modify zoap_client to new CoAP api's

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

mike-scott commented:
V2 posted.  Diff: https://hastebin.com/seyogamofe.cs

Summary of changes:
- Only 1 lwm2m_ctx in the sample since it will use IPv6/IPv4 based on the PEER IP address.
- Removed peer-address, port and endpoint name from the LwM2M context.   They were single use information items. (A lot of code was removed as a result)
- The sample client no longer calls lwm2m_engine_start().  That's moved into the registration client.  This simplifies the sample code even more.
- Addressed comments

mike-scott commented:
I also have a pending DTLS support patch.  I'm not sure if I should tack that on here as well.  Or keep it for a separate PR.

mike-scott commented:
Added a few more commits for the LwM2M cleanup:
- Remove the needless "registered" field.
- Implement timeout checks for the registration client
- Set a default range for the client lifetime setting in Kconfig

mike-scott commented:
Added the last 2 (hopefully) commits to this PR:
- Adding DTLS support for LwM2M client (in the engine and in the sample)

NOTE: This currently is generating errors due to corruption in the decrypted IP header.  This is being addressed in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/1265

mike-scott commented:
@rbtchc This is pretty much everything I have in my local tree.  Feel free to test everything but the DTLS changes.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2174,15 +2175,17 @@ static int handle_request(struct zoap_packet *request,
 void lwm2m_udp_receive(struct lwm2m_ctx *client_ctx, struct net_pkt *pkt,
 		       bool handle_separate_response,
-		       int (*udp_request_handler)(struct zoap_packet *,
+		       int (*udp_request_handler)(struct net_context *net_ctx,
```
Woops, missed this in the changes.

rbtchc commented:
I've applied firmware patches on top of the modifications.
So far so good.

mike-scott commented:
Fixed checkpatch error referencing spaces instead of tabs.  No functional changes.

mike-scott commented:
I've totally re-worked the timeout processing patch "net: lwm2m: implement client state timeout checks" and will probably remove that patch and everything that followed it from this PR.  Then setup a new PR which implements the timeout handling and then another PR with the rest.

Commit:
net: lwm2m: refactor engine to use lwm2m_message structure

Sending an lwm2m message is too difficult.  It requires pending / reply
and other structures to be configured and set by various portions of
the library.  There is also no way to know if a pending message ever
encounters a timeout.

Let's fix this by simplifying the internal LwM2M engine APIs for
handling lwm2m messages:

1. A user calls lwm2m_get_message(lwm2m_ctx) which returns the first
   available lwm2m message from an array of messages per lwm2m_ctx
   (total # of messages per lwm2m_ctx is set via
   CONFIG_LWM2M_ENGINE_MAX_MESSAGES).
2. Next the user sets all of the fields in the message that are
   required (type, code message id, token, etc)
3. Then the user calls lwm2m_init_message(msg).  This initializes the
   underlying zoap_packet, pending and reply structures.
4. Once initialized, the user creates their payload in msg->zpkt.
5. When the user is ready to send, the call lwm2m_send_message(msg).
6. And if for some reason an error occurs at any point, they can free
   up the entire set of structures with: lwm2m_release_message(msg).

Included in the refactoring is a timeout_cb field which can be set in
the LwM2M messages.  If a pending structure ever expires the engine
will call the timeout_cb passing in the msg structure before it's
automatically released.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: add timeout callbacks to registration client

Callbacks are setup for the following states:
- ENGINE_DO_BOOTSTRAP
- ENGINE_DO_REGISTRATION (first registration)
- ENGINE_REGISTRATION_DONE (subsequent client updates)
- ENGINE_DEREGISTER

In most cases, if a timeout occurs the registration engine goes back to
ENGINE_INIT.  The exception is a timeout during client update, which
forces the state machine back to ENGINE_DO_REGISTRATION (skipping a
boostrap).

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
Hello @jukkar and @rbtchc,

I've posted V5, however it's a fairly large diff starting with "net: lwm2m: remove registration client "registered" field".  I decided that I was unhappy with the old implementation of timeout handling.  So, I went and refactored a lot of code so that the internal message sending APIs all use an "lwm2m_message" structure which keeps track of things a lot more cleanly than before.

I also rebased @rbtchc firmware support patches on top.  And I dropped the DTLS patches which will come back once I migrate to the new COAP APIs.

mike-scott commented:
I also rebased this all on top of latest master.

Commit:
net: lwm2m: fix message release for separate reply

In the case of a proxy server translating HTTP -> COAP (known in
the code as "separate reply"), we were leaking lwm2m_message structures.
This was due to pending objects being cleared out during the first ACK,
and no other way of finding a matching message when the follow up packet
was received.  Let's add a second match for reply to make sure we can
find our matching message resources.

NOTE: This change renames find_msg_from_pending() to find_msg() and
makes it a static function as it's only used by the lwm2m_engine.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: RD client start message should be INFO not DEBUG

This is a useful message announcing that the RD client state machine
is starting for a particular connection.  If the log level is set
low so that DBG messages are hidden, then this message goes away.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -156,158 +163,201 @@ do_firmware_transfer_reply_cb(const struct zoap_packet *response,
+	off = parsed_uri.field_data[UF_HOST].off;
+	len = parsed_uri.field_data[UF_HOST].len;
+
+	/* IPv6 is wrapped by brackets */
```
Although these are changes made by me for firmware pull.

line 274~301 & 307 ~ 310 and be removed.
We can pass the host string to net_app_init_udp_client() directly.
Check the patch: https://hastebin.com/qosizocisi.diff

Commit:
net: lwm2m: change registraion update messages to INFO

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: split out lwm2m context initialization

Create an internal function lwm2m_engine_context_init() which sets
the extra packet pools and initializes retransmit work internal to
the LwM2M engine.

This function will be used by firmware pull support which establishes
a new LwM2M context for downloading firmware.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: setup lwm2m ctx for firmware download

Previously, firmware support wasn't initializing the retransmit work
or the extra network packet pools.  Let's fix that.

NOTE: While this fixes the setup of retransmit work, the actual
attempts to re-send packets which are pending is failing.  Needs
another follow-up fix.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: fix packet leak in timeout work

When a packet expires after the pending retries we call
lwm2m_release_message() to free up resources.  This includes
cleanup of the pending structure which calls net_pkt_unref on
the pending packet.  This would normally free up the packet
memory.  However, earlier in the pending processing we add a ref
to the packet so that normal send processing doesn't free up
the memory.   This meant we were leaking packet memory every
time we had an expiration due to timeout.

Let's do an unref prior to calling lwm2m_release_message() to
make sure the packet memory is freed correctly.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: add full block-wise retries for firmware download

UDP packets can be lost in heavy traffic.  Normally we can handle this
with pending packet processing for packets which have not been responded
to with an ACK.  However, due to the time it takes for firmware to
download via CoAP, an extra level of retries should be added.

The process works like this:

Normal pending packets will try to send 3 times fairly quickly.
If that fails, then the timeout callback is called for the firmware
download process.  A retry counter is incremented and the timeout
callback perform a new packet send of the block-wise transfer
packet that is missing, until the retry counter hits a limit (3)
and then the transfer is aborted.

This allows for a longer "outage" to happen during firmware transfer
and the process can still succeed.

NOTE: This patch does not fix a current bug where the pending process
is not re-sending the packets correctly, it only makes the process
more stable with a better chance to work.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: ignore duplicate/older block transfers

During firmware download via block-wise transfer, we can see
packets occaionally get re-transmitted (normal logic in the
pending / retry functions).  However, Both of these packets
end up coming through the reply handler and we should ignore
any block-wise transfer that has a current value less than
where we expect to be.

NOTE: This fixes K64F ethernet transfers where we were getting
too many packets back in the handler.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
Now that 1.10 merge window is opened, I'm going to break this larger PR down and re-submit as smaller sets of patches.  Hopefully we can push these through with good review.

mike-scott commented:
PR w/ the first 3 trivial changes: https://github.com/zephyrproject-rtos/zephyr/pull/1421

Commit:
net: lwm2m: add RD client callbacks for app

Applications may want to be notified when various events
happen in the LwM2M rd client.  Let's implement an event
callback which sends: connect, disconnect and update events.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
@rbtchc I rebased this holding PR on latest master w/ the patches that were in my queue for LwM2M.  This PR does not include my WIP changes for moving to the new CoAP APIs.

NOTE: This includes several new patches which fix packet leaks, firmware download initialization / firmware download retries and add application callbacks from the RD client.

mike-scott commented:
Rebased against current master

rsalveti commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -123,22 +130,29 @@ config LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
 	bool "Firmware Update object pull support"
 	default y
 	depends on LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT
+	select HTTP_PARSER_URL_ONLY
```
This is a problem in case the application also wants to use the HTTP library, as it will force Zephyr to use PARSER_URL_ONLY.

Wonder if it should depend on it instead (HTTP_PARSER || HTTP_PARSER_URL_ONLY), but the problem is that someone needs to set the option.

Commit:
net: coap: parse_option() exit gracefully when no payload

Instead of generating an EINVAL error when there's no payload let's
check for an offset == 0 and exit gracefully.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: move to using the new multi-packet CoAP API

This patch moves from the ZoAP API in subsys/net/lib/zoap to
the CoAP API in subsys/net/lib/coap which handles multiple
fragments for sending / receiving data.

NOTE: This patch moves the LwM2M library over to the CoAP APIs
but there will be a follow-up patch which re-writes the content
formatter reader / writers to use net_pkt APIs for parsing
across multiple net buffers. The current implementation assumes
all of the data will land in 1 buffer.

Samples using the library still need a fairly large NET_BUF_DATA_SIZE
setting. (Example: CONFIG_NET_BUF_DATA_SIZE=384)

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
@rbtchc would you like me to add the initial patch converting the LwM2M library to CoAP APIs?  I still need to re-work the content formatter reader/writer code to use net_pkt buffer API calls to parse, but with this patch you can at least test out basic CoAP API functionality.

rbtchc commented:
@mike-scott Of course.

mike-scott commented:
@rbtchc I've added the series from https://github.com/zephyrproject-rtos/zephyr/pull/1232 by @rveerama1 as well as a WIP patch to convert the LwM2M library to use the new CoAP APIs.

NOTES:
- There is a block in lwm2m_engine.c:2564 where response code is changed to an error response (such as COAP_RESPONSE_CODE_NOT_FOUND).  These are not handled correctly yet.
- At lwm2m_rw_oma_tlv.c:715 there is a FIXME regarding a similar change of response code to COAP_RESPONSE_CODE_CREATED
- I have 2 more patches which add DTLS support to the library but they are still being debugged.
- The underlying content formatter reader/writers need to be re-worked to use net_buf / net_pkt APIs to read and write to the net_bufs.

Let me know what you think.

- Mike

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2385,17 +2451,42 @@ static int handle_request(struct zoap_packet *request,
 	}
 
+	/* Handle blockwise 1 */
+	if (block_ctx) {
```
Are we moving this up because of the removal of zoap_header_set_code()?

mike-scott replied:
Yes, there is a problem..  (thank you for bringing it up)  I need to split this block so that msg->code can be set correctly.  But  coap_append_block1_option() needs to be called after lwm2m_init_message.

mike-scott commented:
I missed 2 commits to the coap framework.  I am busy for tomorrow, but I will add those soon.

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+some adjustments to the configuration may be needed.
+
+The sample will listen for RPL multicast messages and joins with RPL
+Border Router node in DAG network. It exposes few resources through
```
Delete "It exposes few resources through CoAP server role."

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+Border Router node in DAG network. It exposes few resources through
+CoAP server role.
+
+The sample exports the following resources:
```
... and change this to read:

     The sample exports the following resources through the CoAP server role:

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+   /rpl-info/rank
+   /rpl-info/link-metric
+
+These resources allow to toggle led on board (if boards supports) and build
```
Change to:

     These resources allow you to toggle an on-board LED (if the board supports this), and build an

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+Building And Running
+********************
+
+If border router is a Sparrow border router, follow below steps to build and
```
Delete this paragraph here...

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+
+Running Sparrow BR
+==================
+
```
... and put this rewrite here instead:

     Sparrow has its own TLV mechanism (that Zephyr doesn't support)
     to build the topology. A patch is provided in this folder to support building
     the topology with a CoAP based response. If you are using a Sparrow border
     router, follow these build and run steps:

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+   cd products/sparrow-border-router
+   sudo make connect-high PORT=/dev/ttyACM0
+
+Wait until border-router up and running. For web based UI run below python
```
How about:

     Wait until the border router is up and running. For a web-based UI, run this provided python script:

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+   sudo make connect-high PORT=/dev/ttyACM0
+
+Wait until border-router up and running. For web based UI run below python
+script. Unset if your pc is under proxy.
```
Unset what?

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+   cd examples/sparrow
+   ./wsdemoserver.py
+
+Wait until you see "Connected" message on console. Unset proxy in browser
```
How about:

     Wait until you see a "Connected" message on the console.  Open a web browser, unset
     a proxy if you're using one, and open http://127.0.0.1:8000

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+Running RPL node
+================
+
+To build and run RPL node, follow the below steps to build and install
```
follow these steps

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+
+.. code-block:: console
+
+    make pristine and make flash
```
that's not going to work as written.  How about:

     $ make pristine && make flash

dbkinder commented on samples/net/rpl-node/README.rst:
```diff
@@ -0,0 +1,72 @@
+
+    make pristine and make flash
+
+Wait until RPL node joins with Border-Router and update the list in web UI.
```
what do you have to do to "update the list in web UI"?

dbkinder reviewed (changes requested):
suggested edits

Commit:
mbedtls: add lwm2m config for DTLS support

This config is used with samples/net/lwm2m_client.  It varies slightly
from the config-coap.h file used with the COAP samples in that it
has a larger SSL request size.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: coap: fix coap packet length calculation

This calculation reads the length portion of the COAP header to determine
the length of the coap packet.  However, when encrypted via DTLS this
value seems to be getting corrupted.  Let's change this calculation so
that it will work for when DTLS is both enabled and disabled.  Use the
total length of the fragment data and substract back out the headers
to get a correct value.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lib: lwm2m: use correct remote address when DTLS is enabled

The default net_context remote address is scrambled when using a
connection via DTLS.  Instead let's use the dtls context remote.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: reduce LwM2M footprint via lower defaults in Kconfig

Let's use conservative defaults for the LwM2M library to enable
hardware with constrained resources.  Users can increase where
necessary.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

rveerama1 committed:
net: Deprecate ZOAP library

Added deprecated statements to ZOAP library api, structs and enums.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

Commit:
samples: lwm2m: cleanup prj_*.conf files

Support multiple HW with only 2 configs:
1) default operation w/o DTLS support
2) enable DTLS support

NOTE: This also adjusts README.rst

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: lwm2m: reduce footprint of lwm2m configs

LwM2M is intended for constrained devices.  The default samples
settings are quite large by that standard and can be reduced to
reflect actual usage.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
Rebased on master branch and added WIP DTLS support using new CoAP APIs.

This PR is mostly for @rbtchc to keep track of what's in my tree for upstream patches.  But for others who are curious you're welcome to chime in.

NOTE! Still missing from this:
- Some work to clean up lwm2m_msg in the case of an error response back to server in f6fcbc8.
- Update the content formatter reader/writer APIs to use net_pkt parsing to drop the contiguous buffer requirement.

mike-scott commented:
@dbkinder  no need to review this for documentation changes as these patches will flow into normal PRs.

mike-scott commented:
Closing.  These were all merged via different PRs.

mike-scott commented:
I should say.. the rest will be merged in different PRs :)
